### PR TITLE
Add promise support for onRetryAttempt and shouldRetry

### LINKED ISSUE #144: Add support for `Promise` on `onRetryAttempt()`
From this:
https://github.com/googleapis/gaxios/blob/7137a2d82f2905a8042ea97d4dcbc769b3f43a1b/src/retry.ts#L76-L78

To this:
```
if (config.onRetryAttempt) { 
   await config.onRetryAttempt(err); 
} 
```

### PR DESCRIPTION
**Pull request by atjeff**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com//issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Fixes #144 🦕

### CODE REVIEW

**Comment by atjeff** [comment]
I didn't see any harm in allowing promises in `onRetryAttempt`, as we're just notifying the user we're retrying, do you think it should await the completion of the `onRetryAttempt` promise?

File: test/test.retry.ts [hunk 1]
```diff
@@ -219,6 +218,32 @@ describe('🛸 retry & exponential backoff', () => {
     scopes.forEach(s => s.done());
   });
 
+  it('should notify on retry attempts, including promises', async () => {
```
Comment (bcoe) [inline test/test.retry.ts]: I'd be tempted to make this assertion something like:

> accepts async onRetryAttempt handler

**Comment by atjeff** [comment]
Thanks for the review @bcoe! I was seeing a lot of tests failing without the resolve, I'll hopefully find some time to look into why tomorrow. I renamed the test

**Comment by atjeff** [comment]
Found any time @JustinBeckwith ?

**Comment by atjeff** [comment]
Yep, it was me not thinking to await twice. Woops! Pushing that change now.

**Comment by atjeff** [comment]
I'm not sure how I had failing tests, sorry about that. Awaiting shouldRetryFn(err) works fine. Pushed @bcoe

**Comment by atjeff** [comment]
Thanks for taking the time guys!

**Comment by psychowico** [comment]
sorry I dig it back, but from the discussion I expected that retry will wait for both: `onRetryAttempt` and `shouldRetry` functions.
but I clearly see that the code wait only for `shouldRetry` function, but not for `onRetryAttempt`.
https://github.com/googleapis/gaxios/blob/dd3eead940e0bbacb469fd67c7124bfa00d23223/src/retry.ts#L75

do I miss something?

**Pull request merged**